test/bio_pw_callback_test.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "testutil.h"
+
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+
+/* dummy data that needs to be passed to the callback */
+typedef struct CallbackData {
+    char magic;
+    int result;
+} CALLBACK_DATA;
+
+/* constants */
+static char *key_password = "weak_password";
+static int key_password_len = 13;
```

Comment:
Minor/nit: It would be nice to use `size_of()` to calculate size and try to avoid hardcoding's wherever possible.